File: Hash Table/leetcode_03.cpp

// Problem Link ====>> https://leetcode.com/problems/longest-substring-without-repeating-characters/description/

#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        vector<int> lastIndex(256, -1);
        int maxLen = 0;
        int start = 0;
        for (int i = 0; i < s.length(); i++)
        {
            if (lastIndex[s[i]] >= start)
            {
                start = max(start, lastIndex[s[i]] + 1);
            }
            lastIndex[s[i]] = i;
            maxLen = max(maxLen, i - start + 1);
        }
        return maxLen;
    }
};

int main()
{
    string s = "abcabcbb";
    Solution solution;
    int result = solution.lengthOfLongestSubstring(s);
    cout << result << endl;

    return 0;
}


// ===================>>> Approach 1 -

#include <iostream>
#include <vector>
using namespace std;
#include <unordered_set>
class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        int n = s.length();
        int maxLength = 0;
        unordered_set<char> charSet;
        int left = 0;

        for (int right = 0; right < n; right++)
        {
            if (charSet.count(s[right]) == 0)
            {
                charSet.insert(s[right]);
                maxLength = max(maxLength, right - left + 1);
            }
            else
            {
                while (charSet.count(s[right]))
                {
                    charSet.erase(s[left]);
                    left++;
                }
                charSet.insert(s[right]);
            }
        }

        return maxLength;
    }
};

int main()
{
    string s = "abcabcbb";
    Solution solution;

    int maxLength = solution.lengthOfLongestSubstring(s);

    cout << "The length of the longest substring without repeating characters in \"" << s << "\" is: " << maxLength << endl;

    return 0;
}

//===================>>> Approach 2 - Unordered Map

#include <iostream>
#include <vector>
using namespace std;
#include <unordered_map>
class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        int n = s.length();
        int maxLength = 0;
        unordered_map<char, int> charMap;
        int left = 0;

        for (int right = 0; right < n; right++)
        {
            if (charMap.count(s[right]) == 0 || charMap[s[right]] < left)
            {
                charMap[s[right]] = right;
                maxLength = max(maxLength, right - left + 1);
            }
            else
            {
                left = charMap[s[right]] + 1;
                charMap[s[right]] = right;
            }
        }

        return maxLength;
    }
};

int main()
{
    string s = "abcabcbb";
    Solution solution;

    int maxLength = solution.lengthOfLongestSubstring(s);

    cout << "The length of the longest substring without repeating characters in \"" << s << "\" is: " << maxLength << endl;

    return 0;
}

//===================>>> Approach 3 - Integer Array

#include <iostream>
#include <vector>
using namespace std;
class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        int n = s.length();
        int maxLength = 0;
        vector<int> charIndex(128, -1);
        int left = 0;

        for (int right = 0; right < n; right++)
        {
            if (charIndex[s[right]] >= left)
            {
                left = charIndex[s[right]] + 1;
            }
            charIndex[s[right]] = right;
            maxLength = max(maxLength, right - left + 1);
        }

        return maxLength;
    }
};

int main()
{
    string s = "abcabcbb";
    Solution solution;

    int maxLength = solution.lengthOfLongestSubstring(s);

    cout << "The length of the longest substring without repeating characters in \"" << s << "\" is: " << maxLength << endl;

    return 0;
}

File: Hash Table/leetcode_03.py

# Problem Link ====>> https://leetcode.com/problems/longest-substring-without-repeating-characters/description/

class Solution:
  def lengthOfLongestSubstring(self, s):
    max_len = 0
    start = 0
    char_index = {}
    for i in range(len(s)):
      if s[i] in char_index and char_index[s[i]] >= start:
        start = max(char_index[s[i]] + 1, start)
      char_index[s[i]] = i
      max_len = max(max_len, i - start + 1)
    return max_len

def main():
  s = "abcabcbb"
  solution = Solution()
  result = solution.lengthOfLongestSubstring(s)
  print(result)

if __name__ == "__main__":
  main()


# // ===================>>> Approach 1 -
# // Set
def lengthOfLongestSubstring(s):
    charSet = set()
    maxLength = 0
    left = 0

    for right in range(len(s)):
        if s[right] not in charSet:
            charSet.add(s[right])
            maxLength = max(maxLength, right - left + 1)
        else:
            while s[right] in charSet:
                charSet.remove(s[left])
                left += 1
            charSet.add(s[right])

    return maxLength

if __name__ == "__main__":
    s = "abcabcbb"
    maxLength = lengthOfLongestSubstring(s)

    print("The length of the longest substring without repeating characters in", s, "is:", maxLength)


# //= == == == == == == == == == >> > Approach 2 - Unordered Map
def lengthOfLongestSubstring(s):
    char_map = {}
    max_length = 0
    left = 0

    for right in range(len(s)):
        if s[right] not in char_map or char_map[s[right]] < left:
            char_map[s[right]] = right
            max_length = max(max_length, right - left + 1)
        else:
            left = char_map[s[right]] + 1
            char_map[s[right]] = right

    return max_length

if __name__ == "__main__":
    s = "abcabcbb"
    max_length = lengthOfLongestSubstring(s)

    print(f"The length of the longest substring without repeating characters in {s} is: {max_length}")


# //===================>>> Approach 3 - Integer Array
def lengthOfLongestSubstring(s):
    char_index = [None] * 128
    max_length = 0
    left = 0

    for right in range(len(s)):
        if char_index[ord(s[right])] is not None and char_index[ord(s[right])] >= left:
            left = char_index[ord(s[right])] + 1

        char_index[ord(s[right])] = right
        max_length = max(max_length, right - left + 1)

    return max_length


if __name__ == "__main__":
    s = "abcabcbb"
    max_length = lengthOfLongestSubstring(s)

    print("The length of the longest substring without repeating characters in",
          s, "is:", max_length)
